### PR TITLE
[SYCL] Reuse user ptr in buffer & fix default memory allocation

### PR DESCRIPTION
**Pull request by sergey-semenov**

### CODE REVIEW

File: sycl/include/CL/sycl/detail/buffer_impl.hpp [hunk 1]
```diff
@@ -49,17 +50,18 @@ template <typename AllocatorT> class buffer_impl : public SYCLMemObjT {
...
+  buffer_impl(void *HostData, size_t RequiredAlignment, size_t SizeInBytes,
+              const property_list &Props, AllocatorT Allocator = AllocatorT())
       : MSizeInBytes(SizeInBytes), MProps(Props), MAllocator(Allocator) {
 
     if (!HostData)
       return;
 
     set_final_data(reinterpret_cast<char *>(HostData));
-    if (MProps.has_property<property::buffer::use_host_ptr>()) {
+    if (reinterpret_cast<std::uintptr_t>(HostData) % RequiredAlignment == 0 ||
```
Comment (alexeyvoronov) [inline sycl/include/CL/sycl/detail/buffer_impl.hpp]: is it possible there will be a division by zero?

File: sycl/include/CL/sycl/detail/buffer_impl.hpp [hunk 2]
```diff
@@ -49,17 +50,18 @@ template <typename AllocatorT> class buffer_impl : public SYCLMemObjT {
 public:
   buffer_impl(size_t SizeInBytes, const property_list &PropList,
               AllocatorT Allocator = AllocatorT())
-      : buffer_impl((void *)nullptr, SizeInBytes, PropList, Allocator) {}
+      : buffer_impl((void *)nullptr, 0, SizeInBytes, PropList, Allocator) {}
```
Comment (alexeyvoronov) [inline sycl/include/CL/sycl/detail/buffer_impl.hpp]: it is RequiredAlignment?

File: sycl/include/CL/sycl/buffer.hpp [hunk 3]
```diff
@@ -54,14 +54,14 @@ class buffer {
          const property_list &propList = {})
       : Range(bufferRange), MemRange(bufferRange) {
     impl = std::make_shared<detail::buffer_impl<AllocatorT>>(
-        hostData, get_count() * sizeof(T), propList);
+        hostData, sizeof(T), get_count() * sizeof(T), propList);
```
Comment (MrSidims) [inline sycl/include/CL/sycl/buffer.hpp]: I'm not sure that it's the correct way. I can image a situation, where T is user defined structure with sizeof(T) is uneven number. I think it's better to implement a function roundToClosestPowerOfTwo() number and call it here.

Comment (sergey-semenov) [reply sycl/include/CL/sycl/buffer.hpp]: Good point, done.

File: sycl/include/CL/sycl/buffer.hpp [hunk 4]
```diff
@@ -21,7 +21,7 @@ class queue;
 template <int dimensions> class range;
 
 template <typename T, int dimensions = 1,
-          typename AllocatorT = cl::sycl::buffer_allocator>
+          typename AllocatorT = cl::sycl::detail::aligned_allocator<T>>
```
Comment (alexeyvoronov) [inline sycl/include/CL/sycl/buffer.hpp]: align with specification

Comment (MrSidims) [reply sycl/include/CL/sycl/buffer.hpp]: > align with specification

Spec is incorrect at this point. buffer_allocator must be a template class for various of reasons, but it's not. The issue will be created, but for now I would stay with the approach made by Sergey.

Comment (MrSidims) [reply sycl/include/CL/sycl/buffer.hpp]: So, what I mean, I would approve the change unless Alexey hasn't strong objections of doing so.

**Comment by sergey-semenov** [comment]
Resolved conflicts.

File: sycl/include/CL/sycl/detail/aligned_allocator.hpp [hunk 5]
```diff
@@ -46,11 +45,16 @@ class aligned_allocator {
   pointer address(reference Val) const { return &Val; }
   const_pointer address(const_reference Val) { return &Val; }
 
-  // Allocate aligned (to Alignment) memory
+  // Allocate sufficiently aligned memory
   pointer allocate(size_t Size) {
-    Size += Alignment - Size % Alignment;
+    size_t NumBytes = Size * sizeof(value_type);
```
Comment (romanovvlad) [inline sycl/include/CL/sycl/detail/aligned_allocator.hpp]: Suggest adding const to these values.

File: sycl/include/CL/sycl/detail/common.hpp [hunk 6]
```diff
@@ -103,6 +103,9 @@ template <class T> T createSyclObjFromImpl(decltype(T::impl) ImplObj) {
   return T(ImplObj);
 }
 
+// Returns the smallest power of two not less than a
+size_t getNextPowerOfTwo(size_t a);
```
Comment (romanovvlad) [inline sycl/include/CL/sycl/detail/common.hpp]: a -> Var

File: sycl/include/CL/sycl/detail/aligned_allocator.hpp [hunk 7]
```diff
@@ -46,11 +45,16 @@ class aligned_allocator {
...
+    size_t Alignment =
```
Comment (romanovvlad) [inline sycl/include/CL/sycl/detail/aligned_allocator.hpp]: Suggest making Alignment "constexpr" or change the next line from NumBytes | (Alignment - 1)

File: sycl/include/CL/sycl/detail/aligned_allocator.hpp (hunk 5 above)
Comment (sergey-semenov) [reply sycl/include/CL/sycl/detail/aligned_allocator.hpp]: Done

File: sycl/include/CL/sycl/detail/common.hpp (hunk 6 above)
Comment (sergey-semenov) [reply sycl/include/CL/sycl/detail/common.hpp]: Done

File: sycl/include/CL/sycl/detail/aligned_allocator.hpp (hunk 7 above)
Comment (sergey-semenov) [reply sycl/include/CL/sycl/detail/aligned_allocator.hpp]: Replaced modulus with bitwise AND

File: sycl/include/CL/sycl/buffer.hpp (hunk 4 above)
Comment (bader) [reply sycl/include/CL/sycl/buffer.hpp]: No objections from me.

**Pull request merged**